src/include/duckdb/common/types/string.hpp

```diff
@@ -0,0 +1,190 @@
+//===----------------------------------------------------------------------===//
+//                         DuckDB
+//
+// duckdb/common/types/string.hpp
+//
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "duckdb/common/numeric_utils.hpp"
+#include "duckdb/common/typedefs.hpp"
+
+namespace duckdb {
+class String {
+public:
+	// Owning constructors
+	String(std::string str) // NOLINT: allow implicit conversion
+	    : owned_data(std::move(str)), data(owned_data.c_str()), size(owned_data.size()) {
+	}
+
+	String(const char *ptr, const size_t len) : owned_data(ptr, len), data(ptr), size(len) {
+	}
+
+	String(const char *ptr) // NOLINT: allow implicit conversion
+	    : String(ptr, strlen(ptr)) {
+	}
+
+	// Non-owning constructors
+	static String CreateView(const char *ptr, const size_t len) {
+		String str;
+		str.data = ptr;
+		str.size = len;
+		return str;
+	}
+
+	static String CreateView(const char *ptr) {
+		return CreateView(ptr, strlen(ptr));
+	}
+
+public:
+	bool operator==(const String &other) const {
+		const idx_t this_size = GetSize();
+		const idx_t other_size = other.GetSize();
+
+		if (this_size != other_size) {
+			return false;
+		}
+
+		const char *this_data = GetData();
+		const char *other_data = other.GetData();
+
+		for (idx_t i = 0; i < this_size; i++) {
+			if (this_data[i] != other_data[i]) {
+				return false;
+			}
+		}
+		return true;
+	}
+
+	bool operator!=(const String &other) const {
+		return !(*this == other);
+	}
+
+	bool operator<=(const String &other) const {
+		return *this < other || *this == other;
+	}
+
+	bool operator>=(const String &other) const {
+		return !(*this < other);
+	}
+
+	bool operator<(const String &other) const {
+		const char *this_data = GetData();
+		const char *other_data = other.GetData();
+		const idx_t this_size = GetSize();
+		const idx_t other_size = other.GetSize();
+
+		const idx_t length = MinValue<idx_t>(this_size, other_size);
+
+		for (idx_t i = 0; i < length; i++) {
+			if (this_data[i] < other_data[i]) {
+				return true;
+			}
+			if (this_data[i] > other_data[i]) {
+				return false;
+			}
+		}
+		return this_size < other_size;
+	}
+
+	bool operator>(const String &other) const {
+		return !(*this <= other);
+	}
+
+public:
+	idx_t GetSize() const {
+		return size;
+	}
+
+	const char *GetData() const {
+		return data;
+	}
+
+	char operator[](const idx_t pos) const {
+		D_ASSERT(pos < size);
+		return data[pos];
+	}
+
+	bool IsOwned() const {
+		return !owned_data.empty();
+	}
+
+	bool empty() const {
+		return size == 0;
+	}
+
+	const char *c_str() const {
+		return GetData();
```

Comment:
Is the `c_str()` always null-terminated? If so, does the `size` field include the null byte or not?

Reply:
I think https://en.cppreference.com/w/cpp/string/basic_string/c_str implies that in c++11 c_str() is null terminated (but might contain other null values before that).
Basically `s.c_str()[s.size()] == 0` I think it's guaranteed in C++11 for strings.

Unsure what are the implications, either rename since we might want to have a different standard OR using the same convention.